File: shellheim/src/types/recording.ts
// Recording types for terminal session recordings

export interface Recording {
  id: string;
  entry_id: string;
  session_id: string | null;
  name: string;
  description: string | null;
  duration_secs: number | null;
  file_size: number | null;
  terminal_cols: number | null;
  terminal_rows: number | null;
  started_at: string;
  ended_at: string | null;
}

export interface StartRecordingRequest {
  session_id: string;
  name?: string;
  description?: string;
}

export interface StartRecordingResponse {
  recording_id: string;
  session_id: string;
  started_at: string;
}

export interface StopRecordingRequest {
  recording_id: string;
}

export interface StopRecordingResponse {
  recording_id: string;
  duration_secs: number;
  file_size: number;
}

export interface UpdateRecordingRequest {
  name?: string;
  description?: string;
}

// Helper to format duration
export function formatDuration(seconds: number | null): string {
  if (seconds === null || seconds === undefined) return '--:--';
  const mins = Math.floor(seconds / 60);
  const secs = Math.floor(seconds % 60);
  return `${mins}:${secs.toString().padStart(2, '0')}`;
}

// Helper to format file size
export function formatFileSize(bytes: number | null): string {
  if (bytes === null || bytes === undefined) return '--';
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

File: shellheim/src-tauri/src/models/recording.rs
//! Recording model for terminal session recordings

use serde::{Deserialize, Serialize};

/// Recording metadata stored in database
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Recording {
    pub id: String,
    pub account_id: String,
    pub entry_id: String,
    pub session_id: Option<String>,
    pub name: String,
    pub description: Option<String>,
    pub duration_secs: Option<f64>,
    pub file_path: String,
    pub file_size: Option<i64>,
    pub terminal_cols: Option<i32>,
    pub terminal_rows: Option<i32>,
    pub started_at: String,
    pub ended_at: Option<String>,
    pub created_at: String,
}

/// Request to start a new recording
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StartRecordingRequest {
    pub session_id: String,
    pub name: Option<String>,
    pub description: Option<String>,
}

/// Response after starting a recording
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StartRecordingResponse {
    pub recording_id: String,
    pub session_id: String,
    pub started_at: String,
}

/// Request to stop an active recording
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StopRecordingRequest {
    pub recording_id: String,
}

/// Response after stopping a recording
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StopRecordingResponse {
    pub recording_id: String,
    pub duration_secs: f64,
    pub file_size: i64,
}

/// Recording info without file path (for listing)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RecordingInfo {
    pub id: String,
    pub entry_id: String,
    pub session_id: Option<String>,
    pub name: String,
    pub description: Option<String>,
    pub duration_secs: Option<f64>,
    pub file_size: Option<i64>,
    pub terminal_cols: Option<i32>,
    pub terminal_rows: Option<i32>,
    pub started_at: String,
    pub ended_at: Option<String>,
}

impl From<Recording> for RecordingInfo {
    fn from(r: Recording) -> Self {
        Self {
            id: r.id,
            entry_id: r.entry_id,
            session_id: r.session_id,
            name: r.name,
            description: r.description,
            duration_secs: r.duration_secs,
            file_size: r.file_size,
            terminal_cols: r.terminal_cols,
            terminal_rows: r.terminal_rows,
            started_at: r.started_at,
            ended_at: r.ended_at,
        }
    }
}

/// Update recording request
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateRecordingRequest {
    pub name: Option<String>,
    pub description: Option<String>,
}
